mdshare/mdshare.py

```diff
@@ -147,17 +147,56 @@ def fetch(
         max_attempts (int): number of download attempts
         delay (int): delay between attempts in seconds
         blur (float): degree of blur to randomize the delay
-    '''
-    files = search(filename_pattern, repository=repository)
+    """
+    try:
+        from tqdm import tqdm
+        have_tqdm = True
+    except ImportError:
+        have_tqdm = False
+    files = search(filename_pattern, repository=repository, return_sizes=show_progress and have_tqdm)
+
     if len(files) == 0:
         url = '%s/%s' % (repository, filename_pattern)
         print('File not found: ' + url)
         raise HTTPError(url, 404, 'File(s) not found', None, None)
+
+    if have_tqdm and show_progress:
+        callbacks = []
+
+        class wrapper(object):
+            def __init__(self, filename, size):
+                self.t = tqdm(unit='B',
+                              desc='downloading {}'.format(filename),
+                              total=size,
+                              unit_scale=True,
+                              unit_divisor=1024,
+                              )
+
+            def __call__(self, n, block_size, total):
+                downloaded = n * block_size
+                #print(downloaded, total)
+                self.t.update(max(0, downloaded - self.t.n))
+                if downloaded >= total:
+                    increment = int(self.t.total - self.t.n)
+                    if increment > 0:
+                        self.t.update(increment)
+                    self.t.refresh(nolock=True)
+
+        for f, size in files:
+            path = os.path.join(working_directory, f)
```

Comment:
This line fails when testing on my OSX machine:

```Python-Traceback
_________________________________________________________________________________________ test_fetch_npz_file_temp _________________________________________________________________________________________

    def test_fetch_npz_file_temp():
>       files = fetch('mdshare-test-??.txt', working_directory=None)

mdshare/test/test_mdshare.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
mdshare/mdshare.py:186: in fetch
    path = os.path.join(working_directory, f)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = None, p = ('mdshare-test-00.txt',)

    def join(a, *p):
        """Join two or more pathname components, inserting '/' as needed.
        If any component is an absolute path, all previous path components
        will be discarded.  An empty last part will result in a path that
        ends with a separator."""
>       a = os.fspath(a)
E       TypeError: expected str, bytes or os.PathLike object, not NoneType

../../Library/miniconda3/lib/python3.6/posixpath.py:80: TypeError
```

<details>
<summary>conda list</summary>

```
# Name                    Version                   Build  Channel
absl-py                   0.2.2                      py_0    conda-forge
appnope                   0.1.0                    py36_0    conda-forge
asn1crypto                0.24.0                   py36_0    conda-forge
astor                     0.6.2                      py_0    conda-forge
atomicwrites              1.1.5                    py36_0    conda-forge
attrs                     18.1.0                     py_1    conda-forge
backcall                  0.1.0                      py_0    conda-forge
beautifulsoup4            4.6.0                    py36_0    conda-forge
bhmm                      0.6.3                    py36_1    conda-forge
blas                      1.0                         mkl  
bleach                    2.1.3                      py_0    conda-forge
blosc                     1.14.0                        1    conda-forge
bzip2                     1.0.6                         1    conda-forge
ca-certificates           2018.4.16                     0    conda-forge
certifi                   2018.4.16                py36_0    conda-forge
cffi                      1.11.5                   py36_0    conda-forge
chardet                   3.0.4                    py36_0    conda-forge
clangdev                  6.0.0                 default_0    conda-forge
cloog                     0.18.1                        0  
conda                     4.5.8                    py36_0    conda-forge
conda-build               3.10.9                   py36_0    conda-forge
conda-env                 2.6.0                         0    conda-forge
conda-verify              2.0.0                    py36_0    conda-forge
cryptography              2.2.1                    py36_0    conda-forge
cycler                    0.10.0                   py36_0    conda-forge
cython                    0.28.3                   py36_0    conda-forge
decorator                 4.3.0                      py_0    conda-forge
deeptime                  0.1.4.dev15+g1b260f1           <pip>
dill                      0.2.7.1                  py36_0    conda-forge
entrypoints               0.2.3                    py36_1    conda-forge
filelock                  3.0.4                    py36_0    conda-forge
freetype                  2.8.1                         0    conda-forge
gast                      0.2.0                      py_0    conda-forge
glob2                     0.6                        py_0    conda-forge
gmp                       6.1.2                         0    conda-forge
grpcio                    1.12.0           py36hd9629dc_0  
h5py                      2.8.0            py36h470a237_0    conda-forge
hdf5                      1.10.1                        2    conda-forge
html5lib                  1.0.1                      py_0    conda-forge
humanfriendly             4.12.1                     py_0    conda-forge
icu                       58.2                          0    conda-forge
idna                      2.6                      py36_1    conda-forge
intel-openmp              2018.0.3                      0  
ipykernel                 4.8.2                    py36_0    conda-forge
ipython                   6.4.0                    py36_0    conda-forge
ipython_genutils          0.2.0                    py36_0    conda-forge
ipywidgets                7.2.1                    py36_1    conda-forge
isl                       0.19                          0    conda-forge
jedi                      0.12.0                   py36_0    conda-forge
jinja2                    2.10                     py36_0    conda-forge
joblib                    0.11                     py36_0    conda-forge
jpeg                      9b                            2    conda-forge
jsonschema                2.6.0                    py36_1    conda-forge
jupyter                   1.0.0                      py_1    conda-forge
jupyter_client            5.2.3                    py36_0    conda-forge
jupyter_console           5.2.0                    py36_0    conda-forge
jupyter_contrib_core      0.3.3                    py36_1    conda-forge
jupyter_contrib_nbextensions 0.5.0                    py36_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
jupyter_highlight_selected_word 0.2.0                    py36_0    conda-forge
jupyter_latex_envs        1.4.4                    py36_0    conda-forge
jupyter_nbextensions_configurator 0.4.0                    py36_0    conda-forge
kiwisolver                1.0.1                    py36_1    conda-forge
libcxx                    6.0.0                         0    conda-forge
libcxxabi                 4.0.1                hebd6815_0  
libedit                   3.1.20170329                  0    conda-forge
libffi                    3.2.1                         3    conda-forge
libgfortran               3.0.1                h93005f0_2  
libiconv                  1.15                          0    conda-forge
libpng                    1.6.34                        0    conda-forge
libprotobuf               3.5.2                         0    conda-forge
libsodium                 1.0.16                        0    conda-forge
libtiff                   4.0.9                         0    conda-forge
libxml2                   2.9.8                         0    conda-forge
libxslt                   1.1.32                        0    conda-forge
llvm-meta                 6.0.0                         0    conda-forge
llvmdev                   6.0.0                 default_3    conda-forge
llvmlite                  0.23.0                   py36_1    conda-forge
lxml                      4.2.1                    py36_0    conda-forge
markdown                  2.6.11                     py_0    conda-forge
markupsafe                1.0                      py36_0    conda-forge
matplotlib                2.2.2                    py36_1    conda-forge
mdtraj                    1.9.1                    py36_1    conda-forge
mistune                   0.8.3                    py36_1    conda-forge
mkl                       2018.0.3                      1  
mkl_fft                   1.0.2                    py36_0    conda-forge
mkl_random                1.0.1                    py36_0    conda-forge
mock                      2.0.0                    py36_0    conda-forge
more-itertools            4.2.0                      py_0    conda-forge
mpc                       1.1.0                         5    conda-forge
mpfr                      4.0.1                         0    conda-forge
msmtools                  1.2.1                    py36_1    conda-forge
multiprocess              0.70.5                   py36_0    conda-forge
nbconvert                 5.3.1                      py_1    conda-forge
nbformat                  4.4.0                    py36_0    conda-forge
ncurses                   5.9                          10    conda-forge
nglview                   1.1.3                      py_1    conda-forge
ninja                     1.8.2                h2d50403_1    conda-forge
notebook                  5.5.0                    py36_0    conda-forge
numba                     0.38.1                   py36_0    conda-forge
numexpr                   2.6.5                    py36_0    conda-forge
numpy                     1.13.3           py36ha9ae307_4  
olefile                   0.45.1                   py36_0    conda-forge
openblas                  0.2.20                        8    conda-forge
openssl                   1.0.2o                        0    conda-forge
pandas                    0.23.0                   py36_1    conda-forge
pandoc                    2.2.1                hde52d81_0    conda-forge
pandocfilters             1.4.2                    py36_0    conda-forge
parso                     0.2.1                      py_0    conda-forge
pathos                    0.2.1                    py36_1    conda-forge
patsy                     0.5.0                    py36_0    conda-forge
pbr                       4.0.4                      py_0    conda-forge
pexpect                   4.6.0                    py36_0    conda-forge
pickleshare               0.7.4                    py36_0    conda-forge
pillow                    5.1.0                    py36_0    conda-forge
pip                       9.0.3                    py36_0    conda-forge
pkginfo                   1.4.2                    py36_0    conda-forge
pluggy                    0.6.0                      py_0    conda-forge
pox                       0.2.3                    py36_0    conda-forge
ppft                      1.6.4.7.1                py36_0    conda-forge
progress_reporter         1.3.1                    py36_0    conda-forge
prompt_toolkit            1.0.15                   py36_0    conda-forge
protobuf                  3.5.2                    py36_0    conda-forge
psutil                    5.4.6                    py36_0    conda-forge
ptyprocess                0.5.2                    py36_0    conda-forge
py                        1.5.3                      py_0    conda-forge
pycosat                   0.6.3                    py36_0    conda-forge
pycparser                 2.18                     py36_0    conda-forge
pyEMMA                    2.5.2+297.gb0c48dd2           <pip>
pygments                  2.2.0                    py36_0    conda-forge
pyopenssl                 18.0.0                   py36_0    conda-forge
pyparsing                 2.2.0                    py36_0    conda-forge
pyqt                      5.6.0                    py36_5    conda-forge
pysocks                   1.6.8                    py36_1    conda-forge
pytables                  3.4.3                    py36_8    conda-forge
pytest                    3.6.1                    py36_0    conda-forge
python                    3.6.5                         1    conda-forge
python-dateutil           2.7.3                     <pip>
python-dateutil           2.7.3                      py_0    conda-forge
python.app                1.2                      py36_0    conda-forge
pytorch                   0.4.0           py36_cuda0.0_cudnn0.0_1    pytorch
pytz                      2018.4                     py_0    conda-forge
pyyaml                    3.12                     py36_1    conda-forge
pyzmq                     17.0.0                   py36_4    conda-forge
qt                        5.6.2                h9e3eb04_4    conda-forge
qtconsole                 4.3.1                    py36_0    conda-forge
readline                  7.0                           0    conda-forge
requests                  2.18.4                   py36_1    conda-forge
ruamel_yaml               0.15.37          py36h470a237_0    conda-forge
scipy                     1.1.0            py36hcaad992_0  
seaborn                   0.8.1                    py36_0    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                39.2.0                   py36_0    conda-forge
simplegeneric             0.8.1                    py36_0    conda-forge
sip                       4.18                     py36_1    conda-forge
six                       1.11.0                   py36_1    conda-forge
sqlite                    3.20.1                        2    conda-forge
statsmodels               0.9.0                    py36_0    conda-forge
tensorboard               1.8.0                     <pip>
tensorflow                1.8.0                     <pip>
termcolor                 1.1.0                    py36_1    conda-forge
terminado                 0.8.1                    py36_0    conda-forge
testpath                  0.3.1                    py36_0    conda-forge
thermotools               0.2.6            py36h470a237_3    conda-forge
tk                        8.6.7                         0    conda-forge
torchvision               0.2.1                    py36_1    pytorch
tornado                   5.0.2                    py36_0    conda-forge
tqdm                      4.23.4                     py_0    conda-forge
tqdm                      4.23.4                    <pip>
traitlets                 4.3.2                    py36_0    conda-forge
urllib3                   1.22                     py36_0    conda-forge
wcwidth                   0.1.7                    py36_0    conda-forge
webencodings              0.5.1                    py36_0    conda-forge
werkzeug                  0.14.1                     py_0    conda-forge
wheel                     0.31.0                   py36_0    conda-forge
widgetsnbextension        3.2.1                    py36_0    conda-forge
xz                        5.2.3                         0    conda-forge
yaml                      0.1.7                         0    conda-forge
zeromq                    4.2.5                         1    conda-forge
zlib                      1.2.11               h470a237_3    conda-forge
```

</details>